### DISCUSSION TITLE
Cannot test paginated AR collection in view spec - RSpec 6

### DISCUSSION THREAD

**Discussion by cesc1989** [ddnexus/pagy#537] [Q&A] [answered]
**Update: [repo](https://github.com/cesc1989/repro_pagy_rspec_view) with reproduction of the error**

Hi everyone. I'm trying to run this view spec example:

```ruby
# spec/views/expenditures/index.html.erb_spec.rb

require 'rails_helper'

RSpec.describe "expenditures/index" do
  include Pagy::Backend
  include Pagy::Frontend

  it 'displays all current expenditures' do
    expenditures = create_list(:expenditure, 2)

    expenditures = pagy(expenditures)

    assign(:expenditures, expenditures)
    assign(:pagy, expenditures)

    render

    expect(rendered).to match(/Gastos a la fecha/)
  end
end
```

but get this error

```
Failures:

  1) expenditures/index displays all current expenditures
     Failure/Error: expenditures = pagy(expenditures)
     
     NoMethodError:
       undefined method `offset' for [#<Expenditure id: 1, description: "Triple A", amount: 120000, spent_at: "2023-07-09 06:50:34.000000000 +0000", category_id: 1, user_id: 1, created_at: "2023-07-19 04:49:28.794253000 +0000", updated_at: "2023-07-19 04:49:28.794253000 +0000">, #<Expenditure id: 2, description: "Triple A", amount: 120000, spent_at: "2023-07-09 06:50:34.000000000 +0000", category_id: 2, user_id: 2, created_at: "2023-07-19 04:49:28.799816000 +0000", updated_at: "2023-07-19 04:49:28.799816000 +0000">]:Array
     
             collection.offset(pagy.offset).limit(pagy.items)
                       ^^^^^^^
     # /Users/francisco/.gem/ruby/3.1.0/gems/pagy-6.0.4/lib/pagy/backend.rb:29:in `pagy_get_items'
     # /Users/francisco/.gem/ruby/3.1.0/gems/pagy-6.0.4/lib/pagy/backend.rb:14:in `pagy'
     # ./spec/views/expenditures/index.html.erb_spec.rb:15:in `block (2 levels) in <top (required)>'
```

Notice that I `include Pagy::Backend` and `include Pagy::Frontend` or else the error would be about `prev` method.

I've tried to stub the `offset` method to the factory bot list but with no success:

```ruby
allow(expenditures).to receive(:offset)
```

What can I be doing wrong? How can I properly mock `pagy` methods? Any ideas?

Thanks a lot!

**Comment by benkoshy** [comment]
Looks like `create_list` is returning an array?

If that's the case use the `array` extra: https://ddnexus.github.io/pagy/docs/extras/array/

I'll throw in my two cents:

* I cannot fully understand your use case(s) from afar, however, I would suggest, in your controller - do not use the array extra unless you have really cogent reasons to do so. i.e. pagy_array only when testing. see the warnings in the above link.

You may even do something like this if you wish:

```rb
# pagy.rb # config file.
require 'pagy/extras/array' if Rails.env.test?
```

Feel free to re-open the issue if I've missed the mark or you have further questions.

**Reply by cesc1989** [reply]
Thanks for answering.

I forgot to mention that the FactoryBot method `create_list` indeed returns a list.

I gave it a try to the array extra and I got this error.

Looks like it's related to `pagy_bootstrap_nav`

```
1) expenditures/index displays all current expenditures
     Failure/Error: <%== pagy_bootstrap_nav(@pagy) %>
     
     ActionView::Template::Error:
       undefined method `prev' for [#<Pagy:0x000000010ad47f50 @vars={:page=>1, :items=>20, :outset=>0, :size=>[1, 4, 4, 1], :page_param=>:page, :params=>{}, :fragment=>"", :link_extra=>"", :i18n_key=>"pagy.item_name", :cycle=>false, :request_path=>"", :steps=>false, :count=>2}, @count=2, @page=1, @outset=0, @items=20, @last=1, @pages=1, @offset=0, @params={}, @from=1, @to=2, @in=2, @prev=nil, @next=nil>, [#<Expenditure id: 1, description: "Triple A", amount: 120000, spent_at: "2023-07-09 06:50:34.000000000 +0000", category_id: 1, user_id: 1, created_at: "2023-07-19 22:20:23.962513000 +0000", updated_at: "2023-07-19 22:20:23.962513000 +0000">, #<Expenditure id: 2, description: "Triple A", amount: 120000, spent_at: "2023-07-09 06:50:34.000000000 +0000", category_id: 2, user_id: 2, created_at: "2023-07-19 22:20:23.968242000 +0000", updated_at: "2023-07-19 22:20:23.968242000 +0000">]]:Array
     
             p_prev      = pagy.prev
                               ^^^^^
     # /Users/francisco/.gem/ruby/3.1.0/gems/pagy-6.0.4/lib/pagy/frontend.rb:64:in `pagy_link_proc'
     # /Users/francisco/.gem/ruby/3.1.0/gems/pagy-6.0.4/lib/pagy/extras/bootstrap.rb:13:in `pagy_bootstrap_nav'
     # ./app/views/expenditures/index.html.erb:11:in `_app_views_expenditures_index_html_erb__705626454162501274_23540'
```

**Reply by benkoshy** [reply]
Debugging in this manner is very difficult: please help by providing fully reproducible code samples.

Try this:

```rb
require 'rails_helper'

include Pagy::Backend

RSpec.describe "expenditures/index", type: :view do
  it 'displays all current expenditures' do   

    pagy, expenditures = pagy_array(create_list(:expenditure, 2)) 
    assign(:expenditures, expenditures)
    assign(:pagy, pagy)

    render

    expect(rendered).to match(/Description/)
  end
end
```

(I'm not conversant with rspec and what `assigns` does.)

**Reply by cesc1989** [reply]
hey @benkoshy just updated the body of the Q with a link to a repro repo. The instructions in the readme should be clear and accurate 😅 

thanks a lot for the help.

**Comment by ddnexus** [comment, marked as answer]
OK, so there are quite a few things that don't make sense here.

The `pagy*` controller methods accept a collection (which in your case should be an AR scope, not an array) and return an array of 2 items: the pagy object and the paginated collection items, so you are using it improperly. That way the `expenditures` in your test is an array of those 2 items: no wonder it gives errors.

However, you are testing a view, and since you are mocking it rather than running your actual code like in an integration test, you don't need to paginate anything at all with pagy. Pagy is extensively tested with 100% of coverage, so DRY.

In order to test your view you have just to `assign` the local variables, so something like the following should be enough:

```rb
RSpec.describe "expenditures/index" do
  it 'displays all current expenditures' do
    assign(:expenditures, create_list(:expenditure, 2)) # the page array of expenditures
    assign(:pagy, Pagy.new(count: total_count_of_your_collection, **other_pagy_vars)
    render
    expect(rendered).to match(/Gastos a la fecha/)
  end
end
```

If you really want to test your actual code (and how you actually use pagy), then you don't need pagy in your tests: you just need to run your code and test the expected output.

**Reply by cesc1989** [reply]
Thanks for replying. I needed to bring Pagy into the party because the tests started to break asking for the method `prev` which is the one the `pagy` object supports.

```
1) expenditures/index displays all current expenditures
     Failure/Error: <%== pagy_bootstrap_nav(@pagy) %>
     
     ActionView::Template::Error:
       undefined method `prev' for [#<Expenditure id: 1, description: "Triple A", amount: 120000, spent_at: "2023-07-09 06:50:34.000000000 -0500", category_id: 1, user_id: 1, created_at: "2023-07-19 21:18:59.008657000 -0500", updated_at: "2023-07-19 21:18:59.008657000 -0500">, #<Expenditure id: 2, description: "Triple A", amount: 120000, spent_at: "2023-07-09 06:50:34.000000000 -0500", category_id: 2, user_id: 2, created_at: "2023-07-19 21:18:59.014333000 -0500", updated_at: "2023-07-19 21:18:59.014333000 -0500">]:Array
     
             p_prev      = pagy.prev
                               ^^^^^
     # /Users/francisco/.gem/ruby/3.1.0/gems/pagy-6.0.4/lib/pagy/frontend.rb:64:in `pagy_link_proc'
     # /Users/francisco/.gem/ruby/3.1.0/gems/pagy-6.0.4/lib/pagy/extras/bootstrap.rb:13:in `pagy_bootstrap_nav'
     # ./app/views/expenditures/index.html.erb:11:in `_app_views_expenditures_index_html_erb__3385339220587002690_23540'
```

As benkoshy mentioned in other thread, I'd better create a rails app were I can reproduce this in isolation.

**Reply by cesc1989** [reply]
hey. thanks for pointing out another way to proceed. Doing this:

```
assign(:pagy, Pagy.new(count: total_count_of_your_collection, **other_pagy_vars)
```

made the test pass. Thanks a lot!

**Comment by cesc1989** [comment]
Fixed with code proposed by ddnexus. Thank you all.
